src/libsync/ocsprofileconnector.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include "accountfwd.h"
+
+#include <QObject>
+#include <QPixmap>
+#include <QUrl>
+#include <QString>
+
+namespace OCC {
+
+struct HovercardAction
+{
+public:
+    HovercardAction();
+    HovercardAction(QString title, QUrl iconUrl, QUrl link);
+
+    QString _title;
+    QUrl _iconUrl;
+    QPixmap _icon;
+    QUrl _link;
+};
+
+struct Hovercard
+{
+    std::vector<HovercardAction> _actions;
+};
+
+class OcsProfileConnector : public QObject
+{
+    Q_OBJECT
+public:
+    explicit OcsProfileConnector(AccountPtr account, QObject *parent = nullptr);
+
+    void fetchHovercard(const QString &userId);
+    Hovercard hovercard() const;
```

Comment:
return by `const &` because you ever only access the _actions member inside `Hovercard` so copying does not really make sense

Reply:
I access it only with `const auto`, I guess the compiler should be able to optimize that copy away because it is not needed (but I'm not 100% sure). Should I change it regardless?

Reply:
@FlexW Implicit Sharing does not apply to custom classes/structs. You can check if the copy is created to confirm if compiler optimization works or not, but, usually, it is safe to just return a const ref as Matthieu suggested..

Reply:
I know @allexzander, implicit sharing is an implementation by Qt that needs manual work :)
It's just that Kevin tolds me once to never return by reference unless you recognize a performance drop because of that. One thing I like about not returning a reference is that it makes the code simpler.

If I make a simple test program, I can see that in Debug mode it is a little bit more work to return by value. But in release builds, the call gets fully optimized away. See: https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwAcpK4AMngMmAByPgBGmMQgAMyJpAAOqAqETgwe3r7%2BQemZjgJhEdEscQnJtpj2JQxCBEzEBLk%2BfoG19dlNLQRlUbHxSSkKza3t%2BV3j/YMVVaMAlLaoXsTI7BwA9NsA1AAqAJ6pmHtHa8R7aFh7CPGYpHske7SoTOh7hnuYqqyp9AAdCYNABBEHgsyJcLIby3EyJNwAN0wDhICOwEPM0IYsK88MRTnGxEwrAxWLBxK8Dj2QlQbCaxGpBCxAHYrGC9ly9uN0CAQCi0cQEW5wizEtg9iwSJgAProJjNWUIjng1kAERVFNBsKYCgUtPpmDcBn1bNVqS8MVoeGQIAh3L2SNQeE%2BH3Q8sVTFl6TFEDQDHGhoZBCZNPMADYFc0lg7uSZ2XHHVyFEbPUrAZaFAhZTEmMgANYQaNMWOJVWOhOasFJnkEPkC1FEYWIukhsPiyWptjppgQJbXATjWsJivJrkkgjrBg8tMl5XlkcaiGpYh4JGK9i13n8wXNkVtzCM5kY2c9%2BdamvLxc1sFiqVMcL9821w8mvUG7ty1AxbRNy%2Bgo6X6yj%2Bf4OIC7q9j6LqCBAo6jgArFWVZlmOKZpqBTYQegHrzr6sHweyXBPGYTyJMhGqodqjoBkGTBeEQeyRiWTGJOqZ7fr%2BWHASWz43oB3L8Fc/pDgQXwMagkZ7HhMFiSA0lerGnLxomynjjuKBrGJIoigpSr4QQgLSiSUEmAhFgaGZ7E6YidYNq4tAAZWy63gJE6YFOxAzpZ/FVhwKy0JwCG8H4HBaKQqCcG41jWLO6ybExUI8KQBCaP5KwFiACEaPonCSCFaURZwvAKCAOWpWF/mkHAsBIGgLCpHQ8TkJQ9WNfQCTAFwXCkTQtAEPEpUQDEhUxOELRHJwyX1WwggAPIMLQk2VaQWAsIYwDiCt%2BAkg466YKVK2/KiDFbMlYp1IVNoxMQE0eFghWhngLBTVVVAGMACgAGp4JgADuc2nKFyX8IIIhiOwUgyIIigqOoK26MRBhGCgMWWPoeAxKVkArKgqQNIdAC0c1mCVdRNtkLgMO4ngdHooThEMlQjMRRRZAIUx%2BKzGTsww8zDAkxF2BTAh9JMtP5EL5N7aLEwDIzCws7Ycuc3osytPzzOCysqbxZDKUkmdVWBRwwWkKF4WRRwqgBBGhMRpIezAMgyB7N1gJmHsEDRZY1hPLghAvNixF7B4DVNVcwdLLwFVaEsGVZTlJv5ebhVWyVZUpWl8e5RwpOpyt6dZ5VOcosQmTOJIQA

If you like, I can still change the function :)

Reply:
in your case the call will not get inlined because the compiler does not have access to the method body
so this not comparable to your test case
the fact is that `Hovercard` is using stl container and as far as I know the deep copy will happen
a last thing you could check (because I do not have the reference at the moment) is when the item type in the vector is considered a POD data type and the copy is done via `malloc` instead of doing a real copy with constructor calls and so on

Reply:
> in your case the call will not get inlined because the compiler does not have access to the method body

makes sense. Since libsync is a shared library right? Will LTO not help with that?
I will change the signature.

> a last thing you could check (because I do not have the reference at the moment) is when the item type in the vector is considered a POD data type and the copy is done via malloc instead of doing a real copy with constructor calls and so on

still, it would be an expensive copy or?